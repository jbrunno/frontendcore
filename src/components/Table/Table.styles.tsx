import { styled } from '@mui/material/styles';
import TablePagination from '../TablePagination/TablePagination';
import TableRow from '../TableRow/TableRow';

export const TableRowStyled = styled(TableRow)`
  &:hover {
    background-color: ${({ theme }) => theme.palette.action.hover};
  }
`;

export const TablePaginationStyled = styled(TablePagination)`
  border-bottom: ${({ theme }) => theme.fn.spacing(0)};
`;
