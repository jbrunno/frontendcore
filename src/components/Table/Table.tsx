import { Fragment, ReactNode } from 'react';
import { Table as MuiTable } from '@mui/material';
import { Waypoint } from 'react-waypoint';
import {
  TableBody,
  TableContainer,
  TableFooter,
  TableHead,
  TablePaginationProps,
  TableRow,
  TableCell,
} from '../index';
import {
  TableHeader,
  HeadCellKey,
  HeaderArray,
  OrderData,
} from '../TableHeader/TableHeader';
import { TablePaginationStyled, TableRowStyled } from './Table.styles';

export type TableProps<
  BodyData,
  Data extends HeadCellKey,
  ExtraField extends string = '',
> = {
  header?: ReactNode;
  headerData?: HeaderArray<Data, ExtraField>;
  bodyData: Array<BodyData>;
  children: (mapRow: BodyData, index?: number) => ReactNode;
  pagination?:
    | (TablePaginationProps & {
        hasNextPage?: boolean;
      })
    | null;
  orderData?: OrderData<Data, ExtraField>;
  scrollPagination?: boolean;
  stickyHeader?: boolean;
  className?: string;
};

export function Table<
  BodyData,
  Data extends HeadCellKey,
  ExtraField extends string = '',
>({
  header,
  children,
  headerData,
  bodyData,
  pagination,
  orderData,
  scrollPagination,
  stickyHeader = true,
  className,
}: TableProps<BodyData, Data, ExtraField>) {
  const key = (index: number) => index.toString();
  return (
    <TableContainer className={className}>
      <MuiTable stickyHeader={stickyHeader}>
        <TableHead>
          {header}
          {headerData && (
            <TableHeader headerData={headerData} orderData={orderData} />
          )}
        </TableHead>
        <TableBody>
          {bodyData.map((body: BodyData, index) => (
            <Fragment key={key(index)}>
              <TableRowStyled>{children(body, index)}</TableRowStyled>
              <TableRowStyled>
                {scrollPagination &&
                  pagination?.hasNextPage &&
                  index === bodyData.length - 1 && (
                    <TableCell>
                      <Waypoint
                        onEnter={() => {
                          if (pagination?.onPageChange) {
                            pagination.onPageChange(null, pagination.page + 1);
                          }
                        }}
                      />
                    </TableCell>
                  )}
              </TableRowStyled>
            </Fragment>
          ))}
        </TableBody>
        {pagination && !scrollPagination && (
          <TableFooter>
            <TableRow>
              <TablePaginationStyled
                {...pagination}
                rowsPerPageOptions={[5, 10, 25, 50, 100]}
              />
            </TableRow>
          </TableFooter>
        )}
      </MuiTable>
    </TableContainer>
  );
}

export default Table;
