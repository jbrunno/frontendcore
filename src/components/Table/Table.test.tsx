import { render } from '@test-utils/component-testing';
import { Table } from './Table';

test('renders the Table', () => {
  const mock = {
    bodyData: [{ data: 'Mocked Data' }],
    children: (row: Record<'data', string>) => row.data,
  };

  const { container } = render(
    <Table bodyData={mock.bodyData}>{(row) => mock.children(row)}</Table>,
  );
  expect(container).toMatchSnapshot();
});
