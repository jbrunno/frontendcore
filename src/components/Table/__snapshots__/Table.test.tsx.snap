// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the Table 1`] = `
.emotion-0 {
  width: 100%;
  overflow-x: auto;
}

.emotion-1 {
  display: table;
  width: 100%;
  border-collapse: separate;
  border-spacing: 0;
}

.emotion-1 caption {
  font-weight: 400;
  font-size: 0.875rem;
  line-height: 1.125rem;
  letter-spacing: 0.0094rem;
  font-family: Lato,sans-serif;
  padding: 1rem;
  color: #3B3B4199;
  text-align: left;
  caption-side: bottom;
}

.emotion-2 {
  display: table-header-group;
}

.emotion-3 {
  display: table-row-group;
}

.emotion-5 {
  color: inherit;
  display: table-row;
  vertical-align: middle;
  outline: 0;
}

.emotion-5.MuiTableRow-hover:hover {
  background-color: #0000000A;
}

.emotion-5.Mui-selected {
  background-color: rgba(74, 49, 255, 0.08);
}

.emotion-5.Mui-selected:hover {
  background-color: rgba(74, 49, 255, 0.12);
}

.emotion-5:hover {
  background-color: #0000000A;
}

<div>
  <div
    class="MuiTableContainer-root emotion-0"
  >
    <table
      class="MuiTable-root MuiTable-stickyHeader emotion-1"
    >
      <thead
        class="MuiTableHead-root emotion-2"
      />
      <tbody
        class="MuiTableBody-root emotion-3"
      >
        <tr
          class="MuiTableRow-root emotion-4 emotion-5"
        >
          Mocked Data
        </tr>
        <tr
          class="MuiTableRow-root emotion-4 emotion-5"
        />
      </tbody>
    </table>
  </div>
</div>
`;
